### PR TITLE
Update `TIMESTAMP` function implementation and signatures.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Signed-off-by: Yury-Fridlyand <yury.fridlyand@improving.com>

## Description

I referred to [MySQL docs](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html) and tried to reproduce MySQL v.8.0.30 behavior as a reference.

See team review and discussion in https://github.com/Bit-Quill/opensearch-project-sql/pull/139
 
#### Signatures
```
(STRING/DATE/TIME/DATETIME/TIMESTAMP) -> TIMESTAMP
(STRING/DATE/TIME/DATETIME/TIMESTAMP, STRING/DATE/TIME/DATETIME/TIMESTAMP) -> TIMESTAMP
```

#### Notes:
Unlike [MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_timestamp), `TIMESTAMP` function returns a `TIMESTAMP` object (`Instant`), not a `DATETIME`.

#### Test queries:
```SQL
SELECT TIMESTAMP('2020-08-26 13:49:00'), TIMESTAMP('2020-08-26 13:49:00', TIME('12:15:42'));
```

## Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by YANG-DB — DISMISSED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 1]
```diff
@@ -718,15 +718,72 @@ private DefaultFunctionResolver time_to_sec() {
...
+   * STRING/DATE/TIME/DATETIME/TIMESTAMP, STRING/DATE/TIME/DATETIME/TIMESTAMP -> TIMESTAMP
    */
   private DefaultFunctionResolver timestamp() {
     return define(BuiltinFunctionName.TIMESTAMP.getName(),
-        impl(nullMissingHandling(DateTimeFunction::exprTimestamp), TIMESTAMP, STRING),
-        impl(nullMissingHandling(DateTimeFunction::exprTimestamp), TIMESTAMP, DATE),
-        impl(nullMissingHandling(DateTimeFunction::exprTimestamp), TIMESTAMP, DATETIME),
-        impl(nullMissingHandling(DateTimeFunction::exprTimestamp), TIMESTAMP, TIMESTAMP));
+        implWithProperties(nullMissingHandlingWithProperties(DateTimeFunction::exprTimestamp),
+            TIMESTAMP, STRING),
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:728]: does this change depends on https://github.com/opensearch-project/sql/pull/1196 ?

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:728]: No, it uses `ADDTIME` function from #1194

Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:728]: Sorry I meant with implicit cast changes in https://github.com/opensearch-project/sql/pull/1196, we don't need to permute all date time types, right?

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:728]: Oh, I see. Maybe...
Let me rebase this after #1196 merge and test.

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java]: I simplified implementation in 51d82cf. Please, have a look, thanks!

**Review by dai-chen — DISMISSED** [review]
Thanks for the changes!

File: core/src/main/java/org/opensearch/sql/expression/DSL.java [hunk 2]
```diff
@@ -419,7 +419,12 @@ public static FunctionExpression time_to_sec(Expression... expressions) {
   }
 
   public static FunctionExpression timestamp(Expression... expressions) {
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/DSL.java:421]: Why leave this function signature?  There are only about 10 tests to update to remove this.

**Review by acarbonetto — DISMISSED** [review]

File: core/src/main/java/org/opensearch/sql/expression/DSL.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/DSL.java:421]: All functions from this file are used in tests only.

**Review by acarbonetto — APPROVED** [review]

**Review by dai-chen — APPROVED** [review]

**Review by MaxKsyunz — APPROVED** [review]

**Pull request merged**